[mlir][memref] Verify out-of-bounds access for `memref.subview`

matthias-springer opened the pull request:
* Improve the verifier of `memref.subview` to detect out-of-bounds extractions.
* Improve the documentation of `memref.subview` to make clear that out-of-bounds extractions are not allowed. Rewrite examples to use the new `strided<>` notation instead of `affine_map` layout maps. Also remove all unrelated operations (`memref.alloc`) from the examples.
* Fix various test cases where `memref.subview` ops ran out-of-bounds.
* Update canonicalizations patterns to ensure that they do not fold IR if it would generate IR that no longer verifies.

Related discussion on Discourse: https://discourse.llvm.org/t/out-of-bounds-semantics-of-memref-subview/85293

matthias-springer committed:
[mlir][memref] Verify out-of-bounds access for memref.subview

matthias-springer commented on mlir/lib/Interfaces/ViewLikeInterface.cpp:
```diff
@@ -36,6 +36,64 @@ LogicalResult mlir::verifyListOfOperandsOrIntegers(Operation *op,
+    ArrayRef<int64_t> shape, ArrayRef<OpFoldResult> mixedOffsets,
+    ArrayRef<OpFoldResult> mixedSizes, ArrayRef<OpFoldResult> mixedStrides,
+    bool generateErrorMessage) {
+  auto getStaticValues = [](ArrayRef<OpFoldResult> ofrs) {
```
Note: We have something like this in `Dialect/Utils/StaticValueUtils.h`, but it cannot be used due to layering constraints. (Interfaces should not depend on dialects.)

banach-space commented on mlir/test/Conversion/MemRefToLLVM/expand-then-convert-to-llvm.mlir:
```diff
@@ -201,21 +201,20 @@ func.func @subview_const_stride_and_offset(%0 : memref<64x4xf32, strided<[4, 1],
   // CHECK: %[[DESC1:.*]] = llvm.insertvalue %[[BASE_ALIGNED]], %[[DESC0]][1] : !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)>
-  // CHECK: %[[CST_OFF:.*]] = llvm.mlir.constant(8 : index) : i64
-  // CHECK: %[[DESC2:.*]] = llvm.insertvalue %[[CST_OFF]], %[[DESC1]][2] : !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)>
+  // CHECK: %[[C1:.*]] = llvm.mlir.constant(1 : index) : i64
```
[nit] `C1` is very cryptic, `CST_OFF` hints that it's the "constant offset" that's specified together with input argument.

matthias-springer replied:
That was because the same value was used for the offset and the stride, so I just called it `C1`. I changed the example so that they have different values.

banach-space commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
````diff
@@ -1876,118 +1876,65 @@ def SubViewOp : MemRef_OpWithOffsetSizesAndStrides<"subview", [
     ```
 
-    Example 3:
+    Example 4:
````
What happened to Example 3?

matthias-springer replied:
My bad, the numbering got messed up.

But I reduced the number of examples by one. I think the examples cover all interesting cases:

1. static source, static offsets, static sizes, static strides => static result
2. static source, dynamic offsets, dynamic sizes, dynamic strides => dynamic result
3. dynamic source, dynamic offsets, static sizes, dynamic strides => static result size but dynamic result strides
4. rank-reducing case
5. identity case

The main change is that the documentation no longer talks about the affine_map representation, which is outdated. That makes the examples shorter because we don't have to explain which dim/symbol corresponds to which size/stride.

banach-space reviewed (commented):
Code and test changes LGTM, thanks for the improvement!

Just a minor question re the docs - I noticed that the more elaborate example have all been removed. Is that to avoid any examples that are potentially out-of-boundsd?

matthias-springer committed:
address comments

matthias-springer committed:
address comments

matthias-springer commented:
> Just a minor question re the docs - I noticed that the more elaborate example have all been removed. Is that to avoid any examples that are potentially out-of-boundsd?

It's still there. I changed the order of the examples. It's example 3 now. I switched the layout map from the old affine_map representation to the new offset+stride representation. We don't have to talk about symbols/dims anymore, that's why it's shorter.

matthias-springer committed:
address comments

banach-space reviewed (approved):
LGTM, thanks!

I've posted one more nit and a question, but these are nice-to-haves. This is obviously all good, thanks!

**EDIT**
Oh, and thanks for all the answers, that really helped reviewing 🙏🏻

matthias-springer commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -3025,6 +3028,14 @@ LogicalResult SubViewOp::verify() {
+      verifyInBoundsSlice(baseType.getShape(), staticOffsets, staticSizes,
+                          staticStrides, /*generateErrorMessage=*/true);
+  if (!boundsResult.isValid)
+    return getOperation()->emitError(boundsResult.errorMessage);
```
I tried to reuse `produceSubViewErrorMsg` in an earlier version. For that, I have to add another item to the `SliceVerificationResult` enum. `SliceVerificationResult` is currently defined in `BuiltinTypes.h`. I didn't want to include offset-sizes-strides-related behavior in that file because all of that logic is currently in `ViewLikeInterface.h`. So this is mostly a layering issue.

matthias-springer replied:
I think it's worth refactoring this in a follow-up PR.

matthias-springer committed:
address comments
